### ISSUE TITLE
[RMM-7] Retrospective Reflection - Learning (afterModel hook with REINFORCE)

### ISSUE THREAD

**Issue by SkrOYC** [SkrOYC/langchain-middlewares-callbacks-ts#50]
# [RMM-7] Retrospective Reflection - Learning (afterModel hook with REINFORCE)

**Parent Epic:** #43

## Goal

Implement the online learning component of Retrospective Reflection: update reranker weights (W_q, W_m) using REINFORCE algorithm based on LLM citation feedback.

## TDD Checklist

- [ ] **RED**: Write tests for reward computation (+1 cited, -1 not cited, binary correctly)
- [ ] **GREEN**: Implement reward assignment
- [ ] **RED**: Write tests for policy gradient computation (∇_φ log P with known values)
- [ ] **GREEN**: Implement gradient calculation
- [ ] **RED**: Write tests for REINFORCE update (Equation 3 math: Δφ = η·(R-b)·∇_φ log P)
- [ ] **GREEN**: Implement weight update logic
- [ ] **RED**: Write tests for gradient accumulation (batch size 4, partial batch handling)
- [ ] **GREEN**: Implement batching logic
- [ ] **RED**: Write tests for weight persistence (BaseStore save after update)
- [ ] **GREEN**: Implement save to BaseStore
- [ ] **RED**: Write tests for `afterModel` hook integration (full pipeline)
- [ ] **GREEN**: Implement complete hook

### afterModel Hook (`src/middleware/hooks/afterModel.ts`)

```typescript
export const afterModel: AfterModelHook = async (state, runtime) => {
  const citations = runtime.context._citations ?? [];
  const reranker = state._rerankerWeights;
  
  // No citations extracted (malformed or error) → skip RL update
  if (citations.length === 0) {
    return {
      _turnCountInSession: state._turnCountInSession,
    };
  }
  
  // Step 4 of Algorithm 1: REINFORCE Update
  const gradients = computeGradients(citations, reranker);
  
  // Accumulate gradients
  const accumulator = state._gradientAccumulator ?? [];
  accumulator.push(gradients);
  
  let updatedWeights = reranker;
  
  // Apply update when batch is full (4 turns) or at session end
  if (accumulator.length >= 4) {
    updatedWeights = applyGradientUpdate(reranker, accumulator);
    
    // Persist to BaseStore
    const { store, userId } = runtime.context;
    const weightStorage = createWeightStorage(store);
    await weightStorage.saveWeights(userId, updatedWeights);
    
    // Clear accumulator
    accumulator.length = 0;
  }
  
  // Clear citations for next turn
  return {
    _rerankerWeights: updatedWeights,
    _gradientAccumulator: accumulator,
    _citations: [],
    _turnCountInSession: state._turnCountInSession,
  };
};
```

### Reward Computation

```typescript
function computeRewards(citations: CitationRecord[]): number[] {
  // Binary rewards: +1 (cited), -1 (not cited)
  return citations.map(c => c.reward);
}
```

### Policy Gradient Computation

```typescript
interface GradientMatrices {
  queryTransform: number[][];   // ∇_Wq (1536×1536)
  memoryTransform: number[][];  // ∇_Wm (1536×1536)
}

function computeGradients(
  citations: CitationRecord[],
  reranker: RerankerState
): GradientMatrices {
  const η = reranker.config.learningRate;
  const b = reranker.config.baseline;
  
  // Initialize zero gradients
  const gradWq = createZeroMatrix(1536, 1536);
  const gradWm = createZeroMatrix(1536, 1536);
  
  for (const citation of citations) {
    const advantage = citation.reward - b;  // R - b
    
    if (advantage === 0) continue;  // No gradient when reward equals baseline
    
    // Compute ∇_φ log P(M_M | q, M_K; φ)
    // This requires the softmax probabilities from Gumbel sampling
    // Simplified: approximate with adapted embeddings
    
    // For each memory, compute contribution to gradient
    const queryGrad = computeQueryGradient(/* ... */);
    const memoryGrad = computeMemoryGradient(/* ... */);
    
    // Accumulate scaled by advantage
    for (let i = 0; i < 1536; i++) {
      for (let j = 0; j < 1536; j++) {
        gradWq[i][j] += η * advantage * queryGrad[i][j];
        gradWm[i][j] += η * advantage * memoryGrad[i][j];
      }
    }
  }
  
  return { queryTransform: gradWq, memoryTransform: gradWm };
}
```

**Note**: Exact gradient computation requires tracking probabilities from `wrapModelCall`. This is complex - may need to pass probability information through transient state or recompute.

### Simplified Gradient Approach

Given complexity, use simplified policy gradient:

```typescript
function computeSimplifiedGradients(
  citations: CitationRecord[],
  queryEmbedding: number[],
  memoryEmbeddings: number[][],
  reranker: RerankerState
): GradientMatrices {
  // Simplified gradient: outer product of embeddings scaled by advantage
  // ∇_Wq ≈ η·(R-b)·q·q^T
  // ∇_Wm ≈ η·(R-b)·m·m^T
  
  const η = reranker.config.learningRate;
  const b = reranker.config.baseline;
  
  const gradWq = createZeroMatrix(1536, 1536);
  const gradWm = createZeroMatrix(1536, 1536);
  
  for (const citation of citations) {
    const advantage = citation.reward - b;
    
    // Outer product for query gradient
    for (let i = 0; i < 1536; i++) {
      for (let j = 0; j < 1536; j++) {
        gradWq[i][j] += η * advantage * queryEmbedding[i] * queryEmbedding[j];
      }
    }
    
    // Outer product for memory gradient
    const memEmbedding = memoryEmbeddings[citation.turnIndex];
    for (let i = 0; i < 1536; i++) {
      for (let j = 0; j < 1536; j++) {
        gradWm[i][j] += η * advantage * memEmbedding[i] * memEmbedding[j];
      }
    }
  }
  
  return { queryTransform: gradWq, memoryTransform: gradWm };
}
```

### Gradient Accumulation

```typescript
interface GradientAccumulator {
  gradients: GradientMatrices[];
}

function accumulateGradients(
  accumulator: GradientMatrices[],
  newGradients: GradientMatrices,
  batchSize: number
): { accumulator: GradientMatrices[]; shouldApply: boolean } {
  accumulator.push(newGradients);
  
  return {
    accumulator,
    shouldApply: accumulator.length >= batchSize,
  };
}
```

### Weight Update Application

```typescript
function applyGradientUpdate(
  reranker: RerankerState,
  accumulatedGradients: GradientMatrices[]
): RerankerState {
  // Sum all accumulated gradients
  const totalGradWq = createZeroMatrix(1536, 1536);
  const totalGradWm = createZeroMatrix(1536, 1536);
  
  for (const grad of accumulatedGradients) {
    for (let i = 0; i < 1536; i++) {
      for (let j = 0; j < 1536; j++) {
        totalGradWq[i][j] += grad.queryTransform[i][j];
        totalGradWm[i][j] += grad.memoryTransform[i][j];
      }
    }
  }
  
  // Apply updates: W += Σ(ΔW)
  const newWq = reranker.weights.queryTransform.map((row, i) =>
    row.map((val, j) => val + totalGradWq[i][j])
  );
  
  const newWm = reranker.weights.memoryTransform.map((row, i) =>
    row.map((val, j) => val + totalGradWm[i][j])
  );
  
  return {
    ...reranker,
    weights: {
      queryTransform: newWq,
      memoryTransform: newWm,
    },
  };
}
```

### Reward Tests
- [ ] Cited memory gets R = +1
- [ ] Not cited memory gets R = -1
- [ ] [NO_CITE] gives all memories R = -1
- [ ] Binary reward structure correct

### Gradient Tests
- [ ] Gradient shape correct (1536×1536)
- [ ] Advantage scaling works (R - b)
- [ ] Zero advantage produces zero gradient
- [ ] Positive advantage increases selection probability
- [ ] Negative advantage decreases selection probability

### REINFORCE Update Tests
- [ ] Weights update in correct direction
- [ ] Learning rate scales updates
- [ ] Update magnitude reasonable (not exploding)
- [ ] Matrix values remain bounded

### Batch Tests
- [ ] Accumulates up to batch size 4
- [ ] Applies update when batch full
- [ ] Handles partial batch at end
- [ ] Clears accumulator after application

### Persistence Tests
- [ ] Saves weights to BaseStore after update
- [ ] Handles save failure gracefully
- [ ] Updated weights loaded on next invocation

### Hook Tests
- [ ] Full pipeline executes
- [ ] No citations → no update
- [ ] Malformed citations → no update
- [ ] Weights updated and persisted

## Acceptance Criteria

- [ ] All tests pass with Bun:test
- [ ] Weights update correctly via REINFORCE
- [ ] Gradient math verified against Equation 3
- [ ] Batch accumulation working
- [ ] Weights persist to BaseStore
- [ ] Partial batch handling correct
- [ ] Graceful degradation on errors

## Technical Notes

- Gradient computation is approximate (simplified for CPU performance)
- Batch size 4 balances learning stability and responsiveness
- Weights must persist immediately (don't wait for batch if session ends)
- Matrix operations are O(n²) = ~2.3M operations per update (acceptable on CPU)
- Consider weight clipping if values explode (>100 or <-100)
- Monitor weight norms for debugging convergence
